### PR TITLE
feat: http gateway dep tracking

### PR DESCRIPTION
**Pull request by cong-or**
HTTP Gateway with dependency tracking.

### CODE REVIEW

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/event.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+type Path = String;
+
+/// HTTP method e.g GET
+type Method = String;
+
+/// Req body
+type Body = Vec<u8>;
+
+#[derive(Serialize, Deserialize, Debug)]
+/// Msg type for MPSC
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/event.rs:26]: I think we should follow a convention that all Rust code should be together and comments not split it.
```suggestion
#[derive(Serialize, Deserialize, Debug)]
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/event.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+/// HTTP Path
+type Path = String;
+
+/// HTTP method e.g GET
+type Method = String;
+
+/// Req body
+type Body = Vec<u8>;
+
+#[derive(Serialize, Deserialize, Debug)]
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/event.rs:25]:
```suggestion
/// Msg type for MPSC
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    service::{make_service_fn, service_fn},
+};
+use tracing::{error, info};
+
+use super::routing::router;
+
+/// HTTP Gateway port
+const GATEWAY_PORT: u16 = 5000;
+
+#[derive(Debug, Clone)]
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:22]:
```suggestion
/// hostname (node name)
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+/// hostname (node name)
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:23]:
```suggestion
#[derive(Debug, Clone)]
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+pub(crate) struct Hostname(pub String);
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:24]: Is it necessary to have a `pub` visibility for this field ?
Is it possible to hide it ?

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            ]
+            .to_vec(),
+            local_addr: SocketAddr::new([127, 0, 0, 1].into(), GATEWAY_PORT),
+        }
+    }
+}
+
+/// Unique identifier for incoming request
+#[derive(Eq, Hash, PartialEq, Clone, Debug)]
+pub(crate) struct EventUID(pub String);
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:52]: same

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+/// Incoming request client IP
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) struct ClientIPAddr(pub SocketAddr);
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:57]: same

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+/// Has the event been processed
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) struct Processed(pub bool);
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:62]: same

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+/// Is the connection still live
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) struct LiveConnection(pub bool);
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:67]: same

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/mod.rs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+mod event;
+mod gateway_task;
+/// Gateway routing logic
+mod routing;
+
+/// State.
+pub(crate) struct State {
+    /// UID for wasm instance
+    pub instance: Ulid,
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/mod.rs:14]:
```suggestion
    pub(crate) instance: Ulid,
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+/// Everything that hits /api should route to hermes
+const HERMES_ROUTE: &str = "/api";
+
+/// Attempts to wait for a value on this receiver,
+/// returning an error if the corresponding channel has hung up,
+/// or if it waits more than timeout of arbitrary 1 second
+const EVENT_TIMEOUT: u64 = 1;
+
+#[derive(Debug)]
+/// Application name
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:34]:
```suggestion
#[derive(Debug)]
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+/// Everything that hits /api should route to hermes
+const HERMES_ROUTE: &str = "/api";
+
+/// Attempts to wait for a value on this receiver,
+/// returning an error if the corresponding channel has hung up,
+/// or if it waits more than timeout of arbitrary 1 second
+const EVENT_TIMEOUT: u64 = 1;
+
+#[derive(Debug)]
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:33]:
```suggestion
/// Application name
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+/// Application name
+pub(crate) struct AppName(pub String);
+
+#[derive(Debug)]
+/// hostname (node name)
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:38]:
```suggestion
#[derive(Debug)]
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+/// Attempts to wait for a value on this receiver,
+/// returning an error if the corresponding channel has hung up,
+/// or if it waits more than timeout of arbitrary 1 second
+const EVENT_TIMEOUT: u64 = 1;
+
+#[derive(Debug)]
+/// Application name
+pub(crate) struct AppName(pub String);
+
+#[derive(Debug)]
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:37]:
```suggestion
/// hostname (node name)
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    let (app, host) = host
+        .split_once('.')
+        .ok_or(anyhow::anyhow!("Malformed Host header"))?;
+
+    Ok((AppName(app.to_owned()), Hostname(host.to_owned())))
+}
+
+/// Routing by hostname is a mechanism for isolating API services by giving each API its
+/// own hostname; for example, service-a.api.example.com or service-a.example.com.
+pub async fn router(
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:74]:
```suggestion
pub(crate) async fn router(
```

File: hermes/bin/src/runtime_extensions/hermes/kv_store/mod.rs [hunk 16]
```diff
@@ -1,6 +1,6 @@
 //! KV-Store runtime extension implementation.
 
-mod event;
+pub mod event;
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/kv_store/mod.rs:3]:
```suggestion
pub(crate) mod event;
```

**Review by Mr-Leshiy — COMMENTED** [review]
LGTM, except minor suggestions

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs (hunk 5 above)
Comment (cong-or) [reply hermes/bin/src/runtime_extensions/hermes/http_gateway/gateway_task.rs:24]: They are currently shared between multiple files. We can encapsulate later as the connection manager evolves. It's just a placeholder for now.

**Review by Mr-Leshiy — APPROVED** [review]
LGTM

**Pull request merged**